src/IterTools.jl

```
@@ -84,6 +85,27 @@ macro ifsomething(ex)
     end
 end
 
+"""
+    firstrest(xs) -> (f, r)
+
+Return the first element and an iterator of the rest as a tuple.
+
+```jldoctest
+julia> f, r = firstrest(1:3)
+(1, Base.Iterators.Rest{UnitRange{Int64},Int64}(1:3, 1))
+
+julia> collect(r)
+2-element Array{Int64,1}:
+ 2
+ 3
+```
+"""
+function firstrest(xs)
+    f, s = iterate(xs)
```

If `iterate` returns `nothing`, this function should throw the same error as `first` does

---

Hmmm, that's different for each collection type though. I'm tempted to just throw an argument error for empty collections.

---

Best to throw the one that generators throw I think (`ArgumentError: collection must be non-empty`)

---

Alright, we're just throwing an argument error when empty and all other cases seem to work.